### PR TITLE
FORMS-214 - Fixed form labels being lost for some elements

### PR DESCRIPTION
**Pull request by simonmarklar**
when the page is changed, some elements lost their label.
- fixed label rendering code to reuse the previous label or render a new one
- added test for every field type

### CODE REVIEW

File: forms/jqm/element.js [hunk 1]
```diff
@@ -68,9 +71,13 @@ define(function (require) {
...
       if (!this.$label) {
         this.$label = $('<label class="ui-input-text"></label>');
+      }
+
+      if (!this.$el.find(this.$label).length) {
         this.$el.append(this.$label);
       }
-      this.$label.html(this.model.attributes.label || '');
+
+      this.$label.text(this.model.attributes.label || '');
```
Comment (ashishtilara) [inline forms/jqm/element.js]: How would this work with special characters? ie. `&amp;`

File: forms/jqm/element.js [hunk 2]
```diff
@@ -68,9 +71,13 @@ define(function (require) {
     renderLabel: function () {
       if (!this.$label) {
         this.$label = $('<label class="ui-input-text"></label>');
+      }
+
+      if (!this.$el.find(this.$label).length) {
```
Comment (ashishtilara) [inline forms/jqm/element.js:76]: Can't we merge this with previous, so when there is no `$label`, create one append to `$el`

File: forms/jqm/element.js [hunk 3]
```diff
@@ -53,7 +53,10 @@ define(function (require) {
     },
 
     remove: function () {
-      this.$label = null;
+      if (this.$label) {
+        this.$label.remove();
+        this.$label = null;
```
Comment (ashishtilara) [inline forms/jqm/element.js:58]: should this also remove itself from `this.$el`? or `.remove()` does that

Comment (jokeyrhyme) [inline forms/jqm/element.js:58]: `.remove()` detaches the DOM element from the DOM. So the DOM element referred to by `this.$el` will no longer have `this.$label` as a descendant.

File: forms/jqm/element.js (hunk 2 above)
Comment (jokeyrhyme) [inline forms/jqm/element.js:76]: That `if (!this.$label) {` block should:
- create the element
- append it to `this.$el`
- save a reference to it as `this.$label`

So yeah, after the block there shouldn't be any reason to do a jQuery search for the element. Everywhere else in the View, we should just check whether `this.$label` is set and act accordingly.

Comment (simonmarklar) [inline forms/jqm/element.js:76]: no, that is what it used to do and that is why labels on certain elements disappeared on page change.

when a page change happens the view is emptied but the reference to $label is still kept on the backbone view. Therefore on first render we need to check if $label exists, if not, create the label and store a reference to it. I did try and put this into the initialize function but it broke so many tests i left it in the renderLabel funciton

next, on a page change, for certain elements, that label is still attached to the backbone view but not at the DOM. therefore you need to check if its attached to the DOM, and if its not, re-attach it. This was the reason some elements were not rendering the label.

finally render the text of the label.

I believe that there is a more serious issue, im not exactly sure what it is yet, but as this is a highly visible problem a quick fix is warranted.

File: forms/jqm/element.js (hunk 3 above)
Comment (simonmarklar) [inline forms/jqm/element.js:58]: http://api.jquery.com/remove/

Description: Remove the set of matched elements from the DOM.

Similar to .empty(), the .remove() method takes elements out of the DOM. Use .remove() when you want to remove the element itself, as well as everything inside it. In addition to the elements themselves, all bound events and jQuery data associated with the elements are removed. To remove the elements without removing data and events, use .detach() instead.

File: forms/jqm/element.js (hunk 1 above)
Comment (simonmarklar) [inline forms/jqm/element.js]: special characters will be printed as is. The platform sends any special char like ampersands down unencoded, but i can change it back to .html to be sure any special chars are translated correctly

File: forms/jqm/element.js (hunk 2 above)
Comment (jokeyrhyme) [inline forms/jqm/element.js:76]: Another change you have here removes the label from both the DOM and from the View's references. Doesn't that change alone cover this issue?

File: forms/jqm/element.js (hunk 1 above)
Comment (simonmarklar) [inline forms/jqm/element.js]: ok changed it back.

There are 20 other places in forms that use .text to set things so we may want to review them when someone has some spare time

File: forms/jqm/element.js (hunk 2 above)
Comment (simonmarklar) [inline forms/jqm/element.js:76]: You would think so... but no. :(

it was already setting it to null. I just removed it from the DOM before setting it to null, just to be complete. I believe if a view sets up an element and keeps a reference to that element, it should explicitly remove that element and the reference to it.

The strange thing is that the code gets called, but then when you change pages `this.$label` is set to the label that was removed. This is why I think there is something deeper at play - but in the mean time we need to make it look like its working - missing labels easy to spot and make us look bad,

**Comment by jokeyrhyme** [comment]
LGTM :+1:

**Comment by GeetaSajwan** [comment]
LGTM as well

**Comment by jokeyrhyme** [comment]
Has a customer advocate taken a look at the improvements in behaviour here?

**Comment by simonmarklar** [comment]
approved by sam :)
